### ISSUE TITLE
How to set batches per epoch so that each epoch runs on all data only once

### ISSUE THREAD

**Issue by AceMcAwesome77** [MIC-DKFZ/nnDetection#151]
Hi, I am trying to run the test example with 10 train and 10 test scans.  However my progress bar is out of 2600 and takes 2 hours per epoch using GPU, so it must be running on the same training data multiple times during each epoch.  How do I set the train batches per epoch so that each scan is run only once during an epoch?  I found this variable:

trainer_cfg["num_train_batches_per_epoch"]

in nndet/ptmodule/base_module.py but I can't find where this variable is defined; it is not in the nndet/conf/config.yaml file.

Thanks!

**Comment by mibaumgartner** [comment]
It is defined here: https://github.com/MIC-DKFZ/nnDetection/blob/main/nndet/conf/train/v001.yaml

Usually reducing it to 250 with 5 epochs works quite well for the toy data (with a small toy dataset, the larger one needs more epochs).

Note: even after reducing it, the sampler will not make sure to sample cases sequentially since the definition for "sample every case once" is ambiguous in the medical domain (we need to balance foreground and background so not every case can be sampled entirely).

**Comment by AceMcAwesome77** [comment]
Thanks for locating the variable, I see it now.  I am still a little confused by your last statement about not every case being sampled entirely.  Say I have 1000 training scans, and my batch_size=1.  I would not want to set num_train_batches_per_epoch to 1500, because then every epoch some scans would get sampled twice as many times as others.  So I would ideally want to set my num_train_batches_per_epoch to 1000 so each scan is sampled once, between running the validation set.  Are you saying that's not the case here because of the foreground/background properties of this model?  Even if just a subset of each scan is sampled as I go through them, there must be some value to set num_train_batches_per_epoch so that each training scan is sampled exactly once each epoch, right?  In the test example with 10 training scans, I thought that would be num_train_batches_per_epoch=10, please explain if that is wrong.  (that is assuming batch_size=1)

**Comment by mibaumgartner** [comment]
This might seem confusing at first, so here is a more through explanation:

1) Notion of epochs: In the natural image domain an epoch is usually defined as an iteration through the whole dataset. In the medical domain, primarily in 3d scans, this notion might not be suitable though since we can only extract patches from each image and due to varying dataset sizes the number of gradient updates during the training would vary drastically. Thus nnU-Net and also nnDetection, use a constant number of gradient steps during the training (i.e. 60 epochs where each epoch is defined as iterating 2500 batches).

2) Sampling: The sampling mechanism can be divided into two parts:

Sampling foreground: in order to get a proper training going, we need to balance foreground and background (in most cases there is much more background than foreground in the images). Thus, some patches are specifically sampled at foreground positions (which would make the iteration through the dataset in one epoch difficult again). Here, each object of the dataset has the same probability of getting drawn. 

Sampling background: For background patches, it is usually sufficient to sample randomly from the dataset where each image has the same probability of getting drawn.

Due to the sampling mechanism and epoch definition mentioned above, the notion of "sequentially iterating through a dataset" is not present in nnDetection and the number of training batches per epoch can be chosen independently of the number of training cases.
